main.c

```diff
@@ -218,6 +231,49 @@ static const struct wl_keyboard_listener keyboard_listener = {
 	.modifiers = noop,
 };
 
+static void touch_handle_down(void *data, struct wl_touch *touch, uint32_t serial,
+                          uint32_t time, struct wl_surface *surface, int32_t id,
+                          wl_fixed_t x, wl_fixed_t y) {
+	struct slurp_seat *seat = data;
+	if (!seat->already_touched) {
+		seat->already_touched = true;
+		seat->current_output =
+			output_from_surface(seat->state, surface);
+		seat->current_touch_output = seat->current_output;
+		move_pointer(seat, x, y);
+		handle_selection_start(seat);
+	}
+}
+
+static void touch_handle_up(void *data, struct wl_touch *touch, uint32_t serial,
+                        uint32_t time, int32_t id) {
+	struct slurp_seat *seat = data;
+	handle_selection_end(seat);
+	seat->already_touched = false;
+	seat->current_output = NULL;
+	seat->current_touch_output = NULL;
+}
+
+static void touch_handle_motion(void *data, struct wl_touch *touch, uint32_t time,
+                            int32_t id, wl_fixed_t x, wl_fixed_t y) {
+	struct slurp_seat *seat = data;
+	if (seat->current_output == NULL) {
+		seat->current_output = seat->current_touch_output;
+	}
+	move_pointer(seat, x, y);
+	handle_active_selection_motion(seat);
+	seat_set_outputs_dirty(seat);
+}
+
+static const struct wl_touch_listener touch_listener = {
+	.down = touch_handle_down,
+	.up = touch_handle_up,
+	.frame = noop,
+	.motion = touch_handle_motion,
+	.orientation = noop,
+	.shape = noop,
+	.cancel = noop,
```

Comment:
Should we clear our state when this event is received?

Reply:
Added the touch_cancel event fd789e9988ead866633f06917ccf339014714be9.